framework/src/media/audio/audio_manager.c

```diff
@@ -520,74 +521,129 @@ int start_audio_stream_in(void *data, uint32_t frames)
-				srcData.input_frames_used = 0;
-
-				if (src_simple(cur_card->resample.handle, &srcData) != SRC_ERR_NO_ERROR) {
-					meddbg("Fail to resample to %u from %u\n", srcData.desired_sample_rate, srcData.origin_sample_rate);
-					return AUDIO_MANAGER_RESAMPLE_FAIL;
-				}
-
-				ret = get_input_frames_byte_size(srcData.output_frames_gen);
-				if (ret > 0) {
-					memcpy(data + get_input_frames_byte_size(frames_copied), srcData.data_out, ret);
-					frames_copied += srcData.output_frames_gen;
-					frames_used += srcData.input_frames_used;
-
-					medvdbg("Record resampled %d/%d\n", frames_used, frames_total);
-				} else {
-					meddbg("Failed to copy recorded data, get_input_frames_byte_size : %d\n", ret);
-					return AUDIO_MANAGER_RESAMPLE_FAIL;
-				}
-			}
-			ret = frames_copied;
 		} else {
 			ret = pcm_readi(pcm, data, frames);
 		}
 
 		medvdbg("Read %d frames\n", ret);
 		if (ret == -EPIPE) {
 			ret = pcm_prepare(pcm);
+			meddbg("PCM is reprepared\n");
 			if (ret != OK) {
 				meddbg("Fail to pcm_prepare()\n");
 				return AUDIO_MANAGER_XRUN_STATE;
 			}
 		} else if (ret == -EINVAL) {
 			return AUDIO_MANAGER_INVALID_PARAM;
 		} else {
-			return ret;
+			break;
 		}
 	} while ((ret == OK) && (retry--));
 
+	if (cur_card->resample.necessary) {
+		int frames_total = ret;
+		int frames_used = 0;
+		int frames_copied = 0;
+		src_data_t srcData = {0, };
+
+		while (frames_total > frames_used) {
+			srcData.data_in = (void *)(cur_card->resample.readbuffer) + get_input_frames_byte_size(frames_used);
+			srcData.input_frames = frames_total - frames_used;
+			srcData.channels_num = 2;
+			srcData.origin_sample_rate = cur_card->resample.from;
+			srcData.origin_sample_width = SAMPLE_WIDTH_16BITS;
+			srcData.desired_sample_rate = cur_card->resample.to;
+			srcData.desired_sample_width = SAMPLE_WIDTH_16BITS;
+			srcData.input_frames_used = 0;
+			srcData.data_out = data + get_input_frames_byte_size(frames_copied);
+			srcData.out_buf_length = frames_total;
+
+			if (src_simple(cur_card->resample.handle, &srcData) != SRC_ERR_NO_ERROR) {
+				meddbg("Fail to resample to %u from %u\n", srcData.desired_sample_rate, srcData.origin_sample_rate);
+				return AUDIO_MANAGER_RESAMPLE_FAIL;
+			}
+
+			ret = get_input_frames_byte_size(srcData.output_frames_gen);
+			if (ret > 0) {
+				frames_copied += srcData.output_frames_gen;
+				frames_used += srcData.input_frames_used;
+				medvdbg("Record resampled %d/%d\n", frames_used, frames_total);
+			} else {
+				meddbg("Failed to copy recorded data, get_input_frames_byte_size : %d\n", ret);
+				return AUDIO_MANAGER_RESAMPLE_FAIL;
+			}
+		}
+		ret = frames_copied;
+	}
+
 	return ret;
 }
 
+static int resample_stream_out(struct pcm *pcm, audio_card_info_t *cur_card, void *data, uint32_t frames)
+{
+	int ret = 0;
+	int frames_used = 0;
+	int written_frames = 0;
+	int prepare_retry = AUDIO_STREAM_RETRY_COUNT;
+	src_data_t srcData = { 0, };
+
+	while (frames > frames_used) {
```

Comment:
We already know ratio when f/w calls set_audio_stream_out
At that time we can config buffer size based on ratio & allocate resample buffer for stream_out;
see below.
config.period_size = AUDIO_STREAM_VOIP_PERIOD_SIZE * ratio;
resample.buffer = malloc();
Assume that file type is 16k but Target sample is 32k.
And User request 320 frames and then we should resample it to 640 frames.
so, below logic is needed.
while (total_frames > resampled_frames) {
       resample(data_from_upper, resample.buffer);
}
writei(pcm, resample.buffer, frames);

Reply:
Okay, I'll update the writei() with this logic!